pczt/src/common.rs

```diff
@@ -0,0 +1,200 @@
+use std::collections::BTreeMap;
+
+use serde::{Deserialize, Serialize};
+
+use crate::roles::combiner::merge_map;
+
+pub(crate) const FLAG_INPUTS_MODIFIABLE: u8 = 0b0000_0001;
+pub(crate) const FLAG_OUTPUTS_MODIFIABLE: u8 = 0b0000_0010;
+pub(crate) const FLAG_HAS_SIGHASH_SINGLE: u8 = 0b0000_0100;
+
+/// Global fields that are relevant to the transaction as a whole.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub(crate) struct Global {
+    //
+    // Transaction effecting data.
+    //
+    // These are required fields that are part of the final transaction, and are filled in
+    // by the Creator when initializing the PCZT.
+    //
+    pub(crate) tx_version: u32,
+    pub(crate) version_group_id: u32,
+
+    /// The consensus branch ID for the chain in which this transaction will be mined.
+    ///
+    /// Non-optional because this commits to the set of consensus rules that will apply to
+    /// the transaction; differences therein can affect every role.
+    pub(crate) consensus_branch_id: u32,
+
+    /// The transaction locktime to use if no inputs specify a required locktime.
+    ///
+    /// - This is set by the Creator.
+    /// - If omitted, the fallback locktime is assumed to be 0.
+    pub(crate) fallback_lock_time: Option<u32>,
+
+    pub(crate) expiry_height: u32,
```

Comment:
I think the expiry height should be optional. If the length of time needed for construction of a PCZT up to a certain point is significant relative to the block time, then we would want to wait until after that point to set the expiry height.

(We should make the same decision about `expiry_height`, `sapling_anchor`, and `orchard_anchor`, since the same argument about potentially long-lived PCZTs applies to all of them.)

Reply:
The expiry height has implications for how long the inputs that were contributed should be considered to be "locked up" for until they can be available for spending again, if the transaction is not completed. As such, the expiry height needs to be known to and agreed upon by the Constructors, and should just be chosen such that there's enough time to complete the transaction at the outset. I feel less strongly about this than I do about the anchors, which I think _must_ be set by the `Creator`; one downside of requiring the expiry height to be set could be that some users then instead choose a zero expiry height in order to "allow more time" for transaction construction. But if they want to do that, they can just set an expiry height that's far in the future anyway, and at least then there's _some_ bound. Keep this required.